celery_exporter/celery_state.py

```diff
@@ -0,0 +1,169 @@
+            self.uuid = event.get("uuid", CELERY_MISSING_DATA)
+            self.state = TaskState.from_event(state)
+            local_received = event.get("local_received")
+            if self.local_received is None:
+                raise ValueError("Invalid Event: missing local_received")
+            elif not (isinstance(local_received, float)):
+                raise TypeError(
+                    f"Expected type float for local_received, found {type(local_received)}"
+                )
+            self.local_received = local_received
+
+            self.name = event.get("name", CELERY_MISSING_DATA)
+            self.runtime = event.get("runtime")
+
+
+class CeleryState:
+    def __init__(
+        self,
+        event_count: int,
+        task_count: int,
+        queue_by_task: Dict[str, str],
+        tasks: LRUCache[str, Task],
+    ):
+        self.event_count = event_count
+        self.task_count = task_count
+        self.queue_by_task = queue_by_task
+        self.tasks = tasks
+
+    @classmethod
+    def new(cls, max_tasks_in_memory: int) -> "CeleryState":
+        return cls(
+            event_count=0,
+            task_count=0,
+            queue_by_task=defaultdict(),
+            tasks=LRUCache(maxsize=max_tasks_in_memory),
+        )
+
+    def event(self, task: Task):
+        self.event_count += 1
+
+        if self.tasks.get(task.uuid) is None:
+            self.tasks[task.uuid] = task
+
+        if task.state == TaskState.RECEIVED:
+            self.task_count += 1
+
+    def collect(self, event: Dict) -> CollectOutcome:
+        kind: Optional[str] = event.get("type")
+        if kind is None:
+            raise ValueError("Invalid Event: missing type")
+
+        if not (is_task_event(kind)):
+            return (None, None, None, None)
+
+        task = Task()
+        task.update_from_event(event)
+
+        name: str
+        queue: str
+        if task.state in [TaskState.SUCCESS, TaskState.FAILURE, TaskState.REVOKED]:
+            name = self.tasks.pop(task.uuid, task).name
+            queue = self.queue_by_task.get(name, CELERY_MISSING_DATA)
+
+            return (
+                name,
+                task.state.value,
+                task.runtime,
+                queue,
+            )
+        else:
+            self.event(task)
+
+            task_mut = self.tasks[task.uuid]
+            task_mut.state = task.state
+            name = task_mut.name
+            event_queue = event.get("queue")
+            if event_queue is not None:
+                self.queue_by_task[name] = event_queue
+
+            queue = self.queue_by_task.get(name, CELERY_MISSING_DATA)
+            return (name, task.state.value, None, queue)
+
+    def latency(self, event: Dict) -> LatencyOutcome:
+        kind: Optional[str] = event.get("type")
+        if kind is None:
+            raise ValueError("Invalid Event: missing type")
+
+        if not (is_task_event(kind)):
+            return (None, None, None)
```

Comment:
nit: unnecessary parens (might be removed by `black` later?)

Reply:
For later, see https://github.com/grafana/celery-exporter/pull/14#discussion_r1034470686